### PR TITLE
Implement new feature that allow producer to be split into topics for individual components.

### PR DESCRIPTION
**Pull request by tribeiro**
Fix setting up version on conda package.
Pin black version.
Fix upload command in Jenkinsfile.conda.

### CODE REVIEW

**Review by r-owen — COMMENTED** [review]
So far this looks great. I'll look at the component producer after lunch. Meanwhile...some suggestions to consider.

File: python/lsst/ts/salkafka/component_producer_set.py [hunk 1]
```diff
@@ -116,17 +176,113 @@ def signal_handler(self):
...
+                parent = psutil.Process(os.getpid())
+
+                # Child process ids.
+                children = parent.children(recursive=True)
+
+                # Send SIGTERM regardless of what signal was received. This
+                # is the safest signal.
+                for process in children:
+                    print(f"Killing child process {process}.")
+                    process.send_signal(signal.SIGTERM)
```
Comment (r-owen) [inline python/lsst/ts/salkafka/component_producer_set.py:261]: I think you don't need all this. I read https://stackoverflow.com/questions/63729195/how-to-terminate-loop-run-in-executor-with-processpoolexecutor-gracefully and played around a bit. SIGTERM is sent automatically through to the child processes and they will exit gracefully without any need for psutil.

File: python/lsst/ts/salkafka/component_producer_set.py [hunk 2]
```diff
@@ -180,4 +336,160 @@ def make_argument_parser():
...
+            "--validate",
+            dest="validate",
+            required=False,
+            action="store_true",
+            help="Validate input file and exit.",
+        )
+
         return parser
+
+    @staticmethod
```
Comment (r-owen) [inline python/lsst/ts/salkafka/component_producer_set.py:364]: Please make this a classmethod and use cls.schema() for clarity.

File: python/lsst/ts/salkafka/component_producer_set.py [hunk 3]
```diff
@@ -116,17 +176,113 @@ def signal_handler(self):
...
+            print(cls.schema())
+            return 0
+
         if args.wait_for_ack != "all":
             args.wait_for_ack = int(args.wait_for_ack)
 
-        producer_set = ComponentProducerSet(args)
-        await producer_set.done_task
+        if args.file is None:
+            producer_set = ComponentProducerSet(args)
```
Comment (r-owen) [inline python/lsst/ts/salkafka/component_producer_set.py:194]: This should be `cls(args)`

File: python/lsst/ts/salkafka/component_producer_set.py [hunk 4]
```diff
@@ -116,17 +176,113 @@ def signal_handler(self):
...
+
+    @classmethod
+    def create_process(
+        cls, args, queue_len, component, add_ack, commands, events, telemetry
+    ):
+        loop = asyncio.new_event_loop()
+
+        asyncio.set_event_loop(loop)
+
+        producer_set = ComponentProducerSet(
```
Comment (r-owen) [inline python/lsst/ts/salkafka/component_producer_set.py:276]: This should be `producer_set = cls(`

Please look for `ComponentProducerSet` everywhere in this file and:
* In all classmethods replace it with `cls`
* In all staticmethods change that to classmethod and see previous item

**Review by r-owen — CHANGES_REQUESTED** [review]
This looks like a useful addition. As far as changes I feel strongly about:

* Try not using psutil, as per my suggestion and test code I sent you on Slack
* Use classmethods and cls
* The argument parser could use refinement
* You added several new methods and arguments that have no documentation. Please add the missing documentation.

I also suggest adding information about the file mode to doc/index.rst.

File: python/lsst/ts/salkafka/component_producer_set.py (hunk 1 above)
Comment (tribeiro) [reply python/lsst/ts/salkafka/component_producer_set.py:261]: As I mention in slack, and is also in the link above: 

```
If you are generating the SIGINT with Ctrl+C, then the signal is being sent to the entire process group.
```

which is not the case when we run on k8s.

Comment (r-owen) [reply python/lsst/ts/salkafka/component_producer_set.py:261]: Let's talk about this. It's certainly acceptable if you need it, but I'm wondering if k8s can be configured to send SIGTERM in such a way as to render it unnecessary (the equivalent of `kill -TERM -pid` where pid is the process ID).

**Comment by r-owen** [comment]
Another thing I hope we can do better on is the API for `ComponentProducerSet`. The constructor and run methods both take a confusing mix of the container argument `args` plus explicitly named arguments. It was a mistake on my part to pass the arguments as `args`, as it makes the API hard to document. I did it to avoid the large number of Kubernetes configuration arguments. Now that the arguments are so much more complicated and `args` does not have all the information, I think passing in `args` is untenable.

Some thoughts:
* There is a lot of information being passed to the constructor that is only used to call `run`. This requires duplicating long argument lists and the associated documentation. Consider eliminating this by having `amain` call `run` directly, instead of having the constructor do it. That ties in neatly with the next item:
* I think it would be cleaner to have two flavors of `run`: one for the normal mode of producing all topics for one or more producers (e.g. `run_all`), and the other to run a partial set of topics for one component (e.g. `run_partial`). The first would take a list of component names and the second would take a single component name and `commands`, `events`, etc.
* The argument lists would be greatly simplified if the `KafkaInfo` object was passed in already constructed.

**Pull request merged**